hs_access_control/models.py

```diff
@@ -0,0 +1,2349 @@
+            * self is an administrator.
+            * self owns the group.
+            * requesting user is the grantee of resource access.
+        *and* removing the user will not lead to a resource without an owner.
+        There is no provision for revoking lower-level permissions for an owner.
+        If a user is a sole owner and holds other privileges, this call will not remove them.
+        """
+
+        # first check for usage error
+        if not isinstance(this_user, User):
+            raise HSAUsageException("Grantee is not a user")
+        access_user = this_user.uaccess
+
+        if not isinstance(this_resource, GenericResource):
+            raise HSAUsageException("Target is not a resource")
+        access_resource = this_resource.raccess
+
+        if command != CommandCodes.CHECK and command != CommandCodes.DO:
+            raise HSAUsageException("Invalid command code")
+
+        if access_user not in access_resource.holding_users.all():
+            if command == CommandCodes.DO:
+                raise HSAccessException("User does not have access to resource")
+            else:
+                return False
+
+        # User authorization: can make change if
+        #   Admin
+        #   Owner of resource
+        #   Modifying self
+        if self.admin \
+                or access_resource.get_combined_privilege(self.user) == PrivilegeCodes.OWNER \
+                or access_user == self:
+            if UserResourcePrivilege.objects \
+                    .filter(resource=access_resource,
+                            privilege=PrivilegeCodes.OWNER) \
+                    .exclude(user=access_user):
+                # then remove the record.
+                # this does not return an error if the object is not shared with the user
+                if command == CommandCodes.DO:
+                    UserResourcePrivilege.objects.filter(resource=access_resource,
+                                                         user=access_user).delete()
+                return True  # indicate success
+            else:
+                # this prevents one from removing privilege for a user
+                # if any grant is owner. This protects the other grants as well.
+                if command == CommandCodes.DO:
+                    raise HSAccessException("Cannot remove sole resource owner")
+                else:
+                    return False
+        else:
+            if command == CommandCodes.DO:
+                raise HSAccessException("Insufficient privilege to unshare resource")
+            else:
+                return False
+
+    def unshare_resource_with_user(self, this_resource, this_user):
+        """
+        Unshare a resource with a specific user, removing all privilege for that user
+
+        :param this_resource: Resource to unshare
+        :param this_user:  User wich whom to unshare it.
+        :return: True if it is unshared successfully, otherwise false

```

Comment:
@hyi Yes. This one-to-one relationship needs to be changed to 'BaseResource'. Since all other resource types are subclass of 'BaseResource' they all will have access to 'ResourceAccess' model.